cpp/oneapi/dal/backend/primitives/reducer.cpp

```diff
@@ -0,0 +1,221 @@
+reducer_singlepass<Float, BinaryFunctor, UnaryFunctor, Layout>::reducer_singlepass(
+    cl::sycl::queue& queue,
+    BinaryFunctor binary_func,
+    UnaryFunctor unary_func)
+        : binary(binary_func),
+          unary(unary_func),
+          q(queue),
+          max_work_group_size(std::min<std::int64_t>(
+              queue.get_device().template get_info<cl::sycl::info::device::max_work_group_size>(),
+              256)) {}
+
+template <typename Float, typename BinaryFunctor, typename UnaryFunctor, layout Layout>
+cl::sycl::event reducer_singlepass<Float, BinaryFunctor, UnaryFunctor, Layout>::operator()(
+    const Float* input,
+    Float* output,
+    std::int64_t vector_size,
+    std::int64_t n_vectors,
+    std::int64_t work_items_per_group) {
+    const std::int64_t local_buff_size = std::min(this->max_work_group_size, work_items_per_group);
+    auto event                         = this->q.submit([&](cl::sycl::handler& handler) {
+        typedef cl::sycl::
+            accessor<Float, 1, cl::sycl::access::mode::read_write, cl::sycl::access::target::local>
+                local_acc_t;
+        auto partial_reduces = local_acc_t(cl::sycl::range<1>{ local_buff_size }, handler);
+        auto functor_instance =
+            kernel_t{ /*.binary          =*/binary,
+                      /*.unary           =*/unary,
+                      /*.vector_size     =*/static_cast<std::uint32_t>(vector_size),
+                      /*.n_vectors       =*/static_cast<std::uint32_t>(n_vectors),
+                      /*.vectors         =*/input,
+                      /*.reduces         =*/output,
+                      /*.partial_reduces =*/partial_reduces };
+        const cl::sycl::range<2> local_range{ static_cast<size_t>(work_items_per_group), 1 };
+        const cl::sycl::range<2> global_range{ static_cast<size_t>(work_items_per_group),
+                                               static_cast<size_t>(n_vectors) };
+        const cl::sycl::nd_range<2> call_range(global_range, local_range);
+        handler.parallel_for<kernel_t>(call_range, functor_instance);
+    });
+    return event;
+}
+
+template <typename Float, typename BinaryFunctor, typename UnaryFunctor, layout Layout>
+cl::sycl::event reducer_singlepass<Float, BinaryFunctor, UnaryFunctor, Layout>::operator()(
+    array<Float> input,
+    array<Float> output,
+    std::int64_t vector_size,
+    std::int64_t n_vectors,
+    std::int64_t work_items_per_group) {
+    if (work_items_per_group > this->max_work_group_size)
+        throw std::exception();
+    if (input.get_count() < (vector_size * n_vectors))
+        throw std::exception();
+    if (output.get_count() < (vectors_are_rows ? n_vectors : vector_size))
+        throw std::exception();
+    return this->operator()(input.get_data(),
+                            output.get_mutable_data(),
+                            vector_size,
+                            n_vectors,
```

Comment:
Please use consistent naming across lib, to refer to number of  elements we use word `count`. `vector_count` or `row_count`